geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/metadata/CachedLazyMetadata.scala

```diff
@@ -0,0 +1,191 @@
+
+  // cache for our metadata - invalidate every 10 minutes so we keep things current
+  private val metaDataCache =
+    Caffeine.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build(
+      new CacheLoader[(String, String), Option[T]] {
+        override def load(k: (String, String)): Option[T] = {
+          if (tableExists.get) {
+            val (typeName, key) = k
+            scanValue(encodeRow(typeName, key)).flatMap(b => Option(serializer.deserialize(typeName, key, b)))
+          } else {
+            None
+          }
+        }
+      }
+    )
+
+  override def getFeatureTypes: Array[String] = {
+    if (tableExists.get) {
+      val rows = scanRows(None)
+      try {
+        rows.map(r => CachedLazyMetadata.decodeRow(r, typeNameSeparator)._1).toSeq.distinct.toArray
+      } finally {
+        rows.close()
+      }
+    } else {
+      Array.empty
+    }
+  }
+
+  override def read(typeName: String, key: String, cache: Boolean): Option[T] = {
+    if (!cache) {
+      metaDataCache.invalidate((typeName, key))
+    }
+    metaDataCache.get((typeName, key))
+  }
+
+  override def insert(typeName: String, key: String, value: T): Unit = insert(typeName, Map(key -> value))
+
+  override def insert(typeName: String, kvPairs: Map[String, T]): Unit = {
+    ensureTableExists()
+    val inserts = kvPairs.toSeq.map { case (k, v) =>
+      // note: side effect in .map - pre-fetch into the cache
+      metaDataCache.put((typeName, k), Option(v))
+      (encodeRow(typeName, k), serializer.serialize(typeName, k, v))
+    }
+    write(inserts)
+  }
+
+  override def invalidateCache(typeName: String, key: String): Unit = metaDataCache.invalidate((typeName, key))
+
+  override def remove(typeName: String, key: String): Unit = {
+    if (tableExists.get) {
+      delete(encodeRow(typeName, key))
+      // also remove from the cache
+      metaDataCache.invalidate((typeName, key))
+    } else {
+      logger.debug(s"Trying to delete '$typeName:$key' but table does not exist")
+    }
+  }
+
+  override def delete(typeName: String): Unit = {
+    import scala.collection.JavaConversions._
+
+    if (tableExists.get) {
+      val prefix = encodeRow(typeName, "")
+      val rows = scanRows(Some(prefix))
+      try {
+        val all = rows.toBuffer
+        if (all.nonEmpty) {
+          delete(all)
+        }
+      } finally {
+        rows.close()
+      }
+    } else {
+      logger.debug(s"Trying to delete type '$typeName' but table does not exist")
+    }
+    metaDataCache.asMap.keys.foreach(k => if (k._1 == typeName) { metaDataCache.invalidate(k) })
+  }
+
+  def encodeRow(typeName: String, key: String): Array[Byte] =
+    CachedLazyMetadata.encodeRow(typeName, key, typeNameSeparator)
+
+  // checks that the table is already created, and creates it if not
+  private def ensureTableExists(): Unit = tableExists.set(true, false, createTable())
+}
+
+object CachedLazyMetadata {
+
+  private val ESCAPE = "%U+007E%"
```

Comment:
I think your unicode syntax is wrong:
```
scala> println("%U+007E%")
%U+007E%
scala> println("%\u007E%")
%~%
```
For simplicity, it might be better to just use the regular tilde character.